nexus/src/db/lookup.rs

```diff
@@ -0,0 +1,754 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+//! Facilities for looking up API resources from the database
+
+use super::datastore::DataStore;
+use super::identity::Resource;
+use super::model;
+use crate::{
+    authz::{self, AuthorizedResource},
+    context::OpContext,
+    db,
+    db::error::{public_error_from_diesel_pool, ErrorHandler},
+    db::model::Name,
+};
+use async_bb8_diesel::AsyncRunQueryDsl;
+use diesel::{ExpressionMethods, QueryDsl, SelectableHelper};
+use futures::future::BoxFuture;
+use futures::FutureExt;
+use omicron_common::api::external::{LookupResult, LookupType, ResourceType};
+use uuid::Uuid;
+
+// TODO-dap XXX Neither "fetcH" nor "lookup" needs to be a trait now that the
+// macro is defining the impls and the structs
+
+pub trait Fetch {
+    type FetchType;
+    fn fetch(&self) -> BoxFuture<'_, LookupResult<Self::FetchType>>;
+}
+
+// This private module exist solely to implement the "Sealed trait" pattern.
+// This isn't about future-proofing ourselves.  Rather, we don't want to expose
+// an interface that accesses database objects without an authz check.
+mod private {
+    use super::LookupPath;
+    use futures::future::BoxFuture;
+    use omicron_common::api::external::LookupResult;
+
+    // TODO-dap XXX-dap we could probably get rid of this trait and its impls
+    // because we're now calling `lookup_root()` from a macro that can
+    // essentially inline the impl of GetLookupRoot for Key.
+    pub trait GetLookupRoot {
+        fn lookup_root(&self) -> &LookupPath<'_>;
+    }
+
+    pub trait LookupNoauthz {
+        type LookupType;
+        fn lookup(&self) -> BoxFuture<'_, LookupResult<Self::LookupType>>;
+    }
+}
+
+use private::GetLookupRoot;
+use private::LookupNoauthz;
+
+pub trait LookupFor {
+    type LookupType;
+    fn lookup_for(
+        &self,
+        action: authz::Action,
+    ) -> BoxFuture<'_, LookupResult<Self::LookupType>>;
+}
+
+enum Key<'a, P> {
+    Name(P, &'a Name),
+    Id(LookupPath<'a>, Uuid),
+}
```

Comment:
Why does `Name` take the generic param but `Id` takes a `LookupPath`?

Reply:
Yeah, this is kind of a subtle and boring implementation detail, but it's related to your question above.   (And sorry again for the complete lack of comments in this file.)

The builder basically builds a linked list representing the path through the resource tree.  At any given time, the only Rust object that's owned by the caller is the current tail of the list.  And at the end, they're going to do a `fetch()`, which the `opctx` and `datastore` that the caller provided at the beginning (which go into the `LookupPath`).  The question is where are those things stored?  I spent a while playing with different approaches but landed on this: the `LookupPath` remains the head of the list, and when you do the `fetch()`, you walk back to the front to get this information.  (You need to walk back up the list in order to do the lookup anyway.)

If you're doing a lookup by-name, you need the parent container `P` already (to look _it_ up), so `P` appears in `Key::Name`.  At the root of the hierarchy (currently `Organization`, but will probably wind up being `Silo`), `P = LookupPath<'a>`.  In other words, at the root, `Key::Name` looks analogous to `Key::Id`.

If you're doing a lookup by-id, given the assumption I made earlier that this is always happening from the root, then this _is_ the root and it stores the `LookupPath` directly, just like the `Key::Name` variant inside `Organization` does.